src/Fantomas/CodePrinter.fs

```diff
@@ -1322,13 +1322,34 @@ and genExpr astContext synExpr =
             else
                 sepSpace ctx
 
-        atCurrentColumn
-            (genExpr astContext e
-             +> colPre sepSpace sepSpace es (fun e ->
-                    onlyIf (isCompExpr e) (sepSpace +> sepOpenSFixed +> sepSpace)
-                    +> indent
-                    +> appNlnFun e (indentIfNeeded +> genExpr astContext e)
-                    +> unindent))
+        fun ctx ->
+            (ifElseCtx 
+                (exceedsWidth ctx.Config.PageWidth
```

Comment:
Use `expressionFitsOnRestOfLine` instead here.
I would also create separate values for the short and long expression path.
Something like:
```fsharp
fun ctx ->
    let shortExpression = ...

    let longExpression = ...

    expressionFitsOnRestOfLine shortExpression longExpression ctx
```

Reply:
I think you might want to also check for `isCompExpr ` and use the existing expression in that case.

Reply:
I have tried to use `expressionFitsOnRestOfLine` however it directly goes into "short expression mode" and splits even very short apps over multiple lines in some cases. Is that intended?

Reply:
Did you try it in this PR? What you are describing sounds a bit weird.

Reply:
Yes, I have just pushed the version with `expressionFitsOnRestOfLine` so you can see the failing tests where even short expressions are newlined. I assume `shortExpressionWithFallback ` which is used by `expressionFitsOnRestOfLine` is the culprit (see the comment):
```
let private shortExpressionWithFallback (shortExpression: Context -> Context) (fallbackExpression) maxWidth startColumn (ctx: Context) =
    // if the context is already inside a ShortExpression mode and tries to figure out if the expression will go over the page width,
    // we should try the shortExpression in this case.
    match ctx.WriterModel.Mode with
    | ShortExpression infos when (List.exists (fun info -> info.ConfirmedMultiline || info.IsTooLong ctx.Config.PageWidth ctx.Column) infos) ->
        ctx
    | _ -> ....
```